### ISSUE TITLE
does not save file in i3wm

### ISSUE THREAD

**Issue by azzamsa** [phw/peek#177]
thanks for this great application.
I have read others issue when using it in i3wm, I have installed compton. it works fine with enabling peek as floating window even **we can't interact to active window inside recording region**, so I resize peek to give me small space to select the active window and operate it using keyboard.

but, when i hit `stop recording`, and select folder and filename to save, it's appear `rendering`, but no file result. I don't find the output foo.gif in my selected directory.

I compile peek my self using instruction [here](https://github.com/phw/peek#debian)

``` bash
$ peek --version
Peek 1.0.3
````

here is my machine 

``` bash
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 9.1 (stretch)
Release:        9.1
Codename:       stretch
``` 

``` bash
i3 version 4.13 (2016-11-08, branch "4.13") © 2009 Michael Stapelberg and contributors`
``` 

``` bash
$ compton --version
0.1~beta2+20150922
``` 

``` bash
$ ffmpeg --version
ffmpeg version 3.2.5-1 Copyright (c) 2000-2017 the FFmpeg developers
  built with gcc 6.3.0 (Debian 6.3.0-18) 20170516
  configuration: --prefix=/usr --extra-version=1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --disable-stripping --enable-avresample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libebur128 --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmp3lame --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-omx --enable-openal --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libopencv --enable-libx264 --enable-shared
  libavutil      55. 34.101 / 55. 34.101
  libavcodec     57. 64.101 / 57. 64.101
  libavformat    57. 56.101 / 57. 56.101
  libavdevice    57.  1.100 / 57.  1.100
  libavfilter     6. 65.100 /  6. 65.100
  libavresample   3.  1.  0 /  3.  1.  0
  libswscale      4.  2.100 /  4.  2.100
  libswresample   2.  3.100 /  2.  3.100
  libpostproc    54.  1.100 / 54.  1.100`
``` 

``` bash
imagemagick  8:6.9.7.4+dfsg-11+deb9u1
``` 

Please ask me more, on what do you need me to produce an output of my sistem packages or configuration.

Thank you so much.

**Comment by azzamsa** [comment]
Update 1 

**the output produced**, after following comment [here](https://github.com/phw/peek/issues/112)

`<policy domain="resource" name="disk" value="1GiB"/>` 

to 10GiB.

I try to record three times, `5 seconds`, `30 seconds` and `1 minute`. 

- `5 second` recording give me an output with very short duration, it does not record the whole 5 second activity, and it takes 15 minutes to render

- `30 second` recording give me an output with very short duration, it just record the first 1-2 second recording duration, the .pam file in .cache was **2 GB**, and take more than 40 - 1 hour to render.

- `1 minute` recording give me no output, after more than 40 minutes rendering, no output produced, and the pam file was **2.7 GB**

my machine is 
```
CPU: Intel i5-7200U (4) @ 3.1GHz
GPU: NVIDIA GeForce 920MX
Memory: 796MB / 3856MB
```

Note : I do not use my GPU when rendering. but with the same long videos .mkv .webm I have very fast time to render using video editing software.
maybe for this time I have to using screen recorder and converting videos manually to gif using ffmpeg.

**Comment by phw** [comment]
I think this is a duplicate of 
https://github.com/phw/peek/issues/125 and related https://github.com/phw/peek/issues/112 (all coming down to the resource issues with ImageMagick).

But your recording times and resulting processing times are really extreme, far off from what I have experienced myself so far. But your large PAM files suggest a large image size. How large is the screen area you are recording?

**Comment by azzamsa** [comment]
really small, it just "1613x738"

I just run `peek` for 3 second to paste this output here. hope it can help figure out the problems.

``` bash

$ peek
Using screen recorder backend ffmpeg
ffmpeg version 3.2.5-1 Copyright (c) 2000-2017 the FFmpeg developers
  built with gcc 6.3.0 (Debian 6.3.0-18) 20170516
  configuration: --prefix=/usr --extra-version=1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --disable-stripping --enable-avresample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libebur128 --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmp3lame --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-omx --enable-openal --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libopencv --enable-libx264 --enable-shared
  libavutil      55. 34.101 / 55. 34.101
  libavcodec     57. 64.101 / 57. 64.101
  libavformat    57. 56.101 / 57. 56.101
  libavdevice    57.  1.100 / 57.  1.100
  libavfilter     6. 65.100 /  6. 65.100
  libavresample   3.  1.  0 /  3.  1.  0
  libswscale      4.  2.100 /  4.  2.100
  libswresample   2.  3.100 /  2.  3.100
  libpostproc    54.  1.100 / 54.  1.100
[x11grab @ 0x5650c739df00] Stream #0: not enough frames to estimate rate; consider increasing probesize
Input #0, x11grab, from ':0+4,77':
  Duration: N/A, start: 1503741787.976999, bitrate: N/A
    Stream #0:0: Video: rawvideo (BGR[0] / 0x524742), bgr0, 1613x738, 10 fps, 1000k tbr, 1000k tbn, 1000k tbc
Output #0, rawvideo, to '/home/myuser/.cache/peek/peek06DZ5Y.pam':
  Metadata:
    encoder         : Lavf57.56.101
    Stream #0:0: Video: pam, rgb24, 1613x738, q=2-31, 200 kb/s, 10 fps, 10 tbn, 10 tbc
    Metadata:
      encoder         : Lavc57.64.101 pam
Stream mapping:
  Stream #0:0 -> #0:0 (rawvideo (native) -> pam (native))
Press [q] to stop, [?] for help
frame=   22 fps= 11 q=-0.0 Lsize=   76726kB time=00:00:02.20 bitrate=285699.7kbits/s speed=1.13x
video:76726kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 0.000000%
```

it produce `Peek 2017-08-26 17-03.gif` which fine ouput and record all the 2 seconds. 

### second try

I try peek again hour ago, this is the result.

- I try recording my emacs activity on my 1366 x 768 screen, around 30 seconds. hoping get the good result from peek and uploaded here.
- the pam result is 

``` bash
$ du -sh peek63IJ5Y.pam
1.3G    peek63IJ5Y.pam
# and
$ du -sh peek65GZ5Y.pam
1.9G    peek65GZ5Y.pam
```

and become more extreme

``` bash
$ du -hs peek
12G     peek
# after some minutes
$ du -hs peek
8.1G    peek
```
- and it takes more than 25 minutes to render, so i kill the process.

this is my output of  `convert -list resource`
``` bash
$ convert -list resource
Resource limits:
  Width: 16KP
  Height: 16KP
  Area: 128MP
  Memory: 256MiB
  Map: 512MiB
  Disk: 10GiB
  File: 768
  Thread: 4
  Throttle: 0
  Time: unlimited
```
Note : for comparison, I recoreded my screen using kazam for 40 seconds and produce the output 6 mb with .webm and 3 mb of with .mp4, converting such that big files give me 251 mb of first.gif and 220 mb of second.gif with -layers Optimize. and it just take no more than 5-6 second to render with kdenlive or ffmpeg in terminal. ( I am writing this to assuere that my machine is healty).

**Comment by phw** [comment]
> ... converting such that big files give me 251 mb of first.gif and 220 mb of second.gif with -layers Optimize. and it just take no more than 5-6 second to render with kdenlive or ffmpeg in terminal. ( I am writing this to assuere that my machine is healty).

Can you explain more what you did, a bit confused by the use of `-layers Optimize` and the statement that it took only seconds with ffmpeg? Did you use ffmpeg or ImageMagick to do the conversion? Trying to understand the difference here. My experience is that as soon as you use convert with `-layers optimize` the resource usage blows up. For reference Peek just runs (variables just added to make it clear what is used there):

    convert -debug None -set delay $framedelay -layers Optimize -define registry:temporary-path=$sometempdir $inputfile $outputfile

**Comment by phw** [comment]
> really small, it just "1613x738"

Unrelated to the issue here, but recording this area as GIF is not really small, especially when doing more than just a few seconds. IMHO you are better served with setting WebM or MP4 as output format here. You can also experiment with the framerate, but Peek's default framerate of 8fps is already pretty low and tuned towards GIF recording.

**Comment by azzamsa** [comment]
>Can you explain more what you did, a bit confused by the use of -layers Optimize and the statement that it took only seconds with ffmpeg? Did you use ffmpeg or ImageMagick to do the conversion? Trying to understand the difference here. My experience is that as soon as you use convert with -layers optimize the resource usage blows up. For reference Peek just runs (variables just added to make it clear what is used there)

I am sorry, I have made you confused. I mean in `Note` section is the step I use when I don't use peek. because peek takes loot of my CPUS and RAM. so I try to record my screen using Kazam Scren recoder and convert the .webm/.mp4 result manually.

> Unrelated to the issue here, but recording this area as GIF is not really small, especially when doing more than just a few seconds. IMHO you are better served with setting WebM or MP4 as output format here. You can also experiment with the framerate, but Peek's default framerate of 8fps is already pretty low and tuned towards GIF recording.

yes, Yesterday I am looking for the standard duration and area for gif recording, any suggestion ?
I will try to make gif with `Peek` with your preferred timing duration and area size, so i will report the result here

**Comment by azzamsa** [comment]
I just want to share my experience with [Silentcast](https://github.com/colinkeenan/silentcast) , maybe this can give a benefit.

Silencast does not natively support i3wm, but this is the workaround. [Does't work in i3wm #49](https://github.com/colinkeenan/silentcast/issues/49). the cool piece is [temptoanim](https://raw.githubusercontent.com/colinkeenan/silentcast/0d9c613e6a9843188b5c297655c2aa7d76a4e369/temptoanim)  it convert the temp.mkv result from scren record to gif with relatively fast and does not eat lot of my machine resources.

this is my GIF result.

I just want to share, I hope this can help this issue.

![emacs-gif](https://user-images.githubusercontent.com/17734314/29855362-7e40cdd2-8d75-11e7-8476-b0701ea5b352.gif)

**Comment by phw** [comment]
The reason I am not recording to webm/mp4 and convert this to gif are:

1. As both formats are lossy and can cause color artifacts, this sometimes lead to even more color problems when doing the GIF conversion. Got better results when converting directly the lossless input (see https://github.com/phw/peek/issues/116).

2. Whenever I tried to convert webm/mp4 with convert directly, it first converted the input file into a temporary .pam file and then ran the GIF conversion on that. So Peek now records directly to .pam and feeds that to convert, avoiding the additional conversion step (which took quite some time and extra disk space)

What I see most other tools are doing is saving the frames as separate PNG files. Maybe this is faster, have to test. Could even save some space for the temporary files. Downside is I cannot easily do this with the gnome-shell backend, as I cannot control the final output sink of the gstreamer pipeline used and it expects to output a single file.

**Comment by phw** [comment]
Had a quick look at silentcast: It first records lossless video file as ffvhuff in a Matroska container, converts this to PNG files and then runs ImageMagick on the PNG files (with the same `-layers optimize` option turned on).

I had always tried avoiding any additional conversion step to an intermediate format (PNG in this case), but will try and compare. If this turns out to be more resource and / or time efficient I will use this. The important things to measure are:

- Total conversion time
- Maximum memory use
- Maximum temporary disk space use

**Comment by phw** [comment]
Some preliminary results: I have two files `input.pam` (274.6). and `input.mkv` (85.7 MB). The PAM file is a rawvideo file as used currently by Peek, the input.mkv is a ffvhuff encoded lossless file. Both have  identical content (actually I recorded the MKV file with a patched Peek and converted the PAM file from it with ffmpeg). The recording is 10 seconds long at 8fps and a size of 1009 x 1055 px.

I convert the PAM file to GIF similar how Peek does it currently using the following script:

    convert -set delay 12.5 \
      -layers Optimize \
      -define registry:temporary-path=$(dirname $0)/tmp \
      input.pam output-from-pam.gif

For the MKV file I convert it to PNG first and then convert the PNGs to GIF (similar to how silentcast does it):

    ffmpeg -i input.mkv -r 8 png/f-%03d.png
    convert -adjoin -delay 12.5 } \
      -layers Optimize \
      png/f-[0-9][0-9][0-9].png output-from-mkv.gif

The result is in both cases a 6.6 MB sized GIF, the output of both conversions is binary identical. I measured the execution time and RAM usage using the "time" utility:

    Command being timed: "./convert-mkv.sh"
    User time (seconds): 47.03
    System time (seconds): 3.33
    Percent of CPU this job got: 145%
    Elapsed (wall clock) time (h:mm:ss or m:ss): 0:34.56
    Average shared text size (kbytes): 0
    Average unshared data size (kbytes): 0
    Average stack size (kbytes): 0
    Average total size (kbytes): 0
    Maximum resident set size (kbytes): 2874384
    Average resident set size (kbytes): 0
    Major (requiring I/O) page faults: 0
    Minor (reclaiming a frame) page faults: 1832389
    Voluntary context switches: 1888
    Involuntary context switches: 13678
    Swaps: 0
    File system inputs: 8
    File system outputs: 37936
    Socket messages sent: 0
    Socket messages received: 0
    Signals delivered: 0
    Page size (bytes): 4096
    Exit status: 0

    Command being timed: "./convert-pam.sh"
    User time (seconds): 39.59
    System time (seconds): 3.21
    Percent of CPU this job got: 134%
    Elapsed (wall clock) time (h:mm:ss or m:ss): 0:31.73
    Average shared text size (kbytes): 0
    Average unshared data size (kbytes): 0
    Average stack size (kbytes): 0
    Average total size (kbytes): 0
    Maximum resident set size (kbytes): 2873844
    Average resident set size (kbytes): 0
    Major (requiring I/O) page faults: 1
    Minor (reclaiming a frame) page faults: 1763222
    Voluntary context switches: 1582
    Involuntary context switches: 9802
    Swaps: 0
    File system inputs: 24
    File system outputs: 12944
    Socket messages sent: 0
    Socket messages received: 0
    Signals delivered: 0
    Page size (bytes): 4096
    Exit status: 0

So in summary converting to PNG first was about 8% slower, both methods used about the same amount of maximum RAM (2.74 GB). No win here for the PNG method.

The temporary files created by Peek are much smaller for PNG method:

- PAM file: **274.6 MB**
- MKV file + PNGs: 85.7 MB + 12.6 MB = **98,3 MB**

Did not measure any temporary files created by ffmpeg or convert processes, if any, but I guess this is about the same in both cases.

I have to test this with a larger file, one of the category that really makes trouble. But so far I don't see much difference. The longer processing time could be an acceptable trade off for lower temp file sizes, as those tend to get very big at the moment.

**Comment by phw** [comment]
For reference this is the GIF result from the above test:

![output-from-mkv](https://user-images.githubusercontent.com/29852/30540249-a1a161d2-9c76-11e7-9caf-18187f443451.gif)

**Comment by phw** [comment]
I am not sure I will continue on this issue, my tests brought no real benefit of the silentcast method.

But please try the latest release of Peek (1.1.0). Peek now handles the memory limits of ImageMagick. In your output above one can see that you have a memory limit of 256MB by default. This is really low for convert, and it will fall back of using disk instead, which is way slower and could explain your long rendering times. The new Peek will allow convert to use 90% of free RAM (at time of invoking convert).

**Comment by azzamsa** [comment]
I will try everything above as you suggest, and report the result here, soon.

Thanks lot for your contribution to community :)

**Comment by phw** [comment]
I just pushed changes that allow using ffmpeg directly instead of ImageMagick to convert the video to GIF. In my first tests this runs far less resource intensive and faster and produces sometimes even smaller output files. Could you give latest master branch a go and let me know if this works for you?

In order to use it you will have for now to start Peek with `PEEK_POSTPROCESSOR=ffmpeg` set, e.g.

    PEEK_POSTPROCESSOR=ffmpeg peek

If ImageMagick is out of the equation I can also use compressed temporary video files (e.g. ffvhuff as in my tests above), greatly reducing the temporary space required.

**Comment by phw** [comment]
I close this, as latest Peek from git now uses either ffmpeg or (if installed) gifski, both tools don't use that exorbitant amount of RAM. Also the temporary files are now much smaller. Overall I think peek now does solve this even better than silentcast.

**Comment by azzamsa** [comment]
Sorry, I have so many task at hand, and I can report peek here. 
Thank you very much for your contribution to community.

I still report peek if I have time in the following weeks.
